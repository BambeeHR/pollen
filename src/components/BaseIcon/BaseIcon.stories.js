import BaseIcon, { Icons } from './BaseIcon.vue';
import '../../assets/tailwind.css';

export default {
  title: 'BaseIcon',
  component: BaseIcon,
};

export const Gallery = () => ({
  components: { BaseIcon },
  data() {
    return {
      icons: Object.values(Icons),
    };
  },
  template: `
    <div class="flex flex-wrap">
      <div v-for="icon in icons" :key="icon" :style="{ margin: '12px', textAlign: 'center' }">
        <BaseIcon :icon="icon" />
        <div :style="{ fontSize: '12px' }">{{ icon }}</div>
      </div>
    </div>
  `,
});
