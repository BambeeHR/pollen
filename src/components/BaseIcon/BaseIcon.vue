<template functional>
  <component
    :is="$options.getComponent(props.icon)"
    :class="[
      data.class,
      data.staticClass,
      { 'base-icon--custom': $options.isCustomIcon(props.icon) },
    ]"
    :icon="$options.getIcon(props.icon)"
    class="base-icon"
    v-bind="$options.getAttributes({ data, parent })"
    v-on="listeners"
  />
</template>

<script>
import camelCase from 'lodash/camelCase';
import snakeCase from 'lodash/snakeCase';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';

import {
  faAnalytics as analytics,
  faBallotCheck as ballotCheckSolid,
  faClipboardCheck as clipboardCheckSolid,
  faCreditCard as creditCardSolid,
  faEar as ear,
  faTachometer as tachometerSolid,
  faUser as userSolid,
  faArrowLeft as arrowLeft,
  faArrowRight as arrowRight,
  faBan as ban,
  faBell as bell,
  faBuilding as building,
  faCabinetFiling as cabinetFilingSolid,
  faCalendar as calendar,
  faCheck as check,
  faCheckCircle as checkCircleSolid,
  faClock as clock,
  faCloudUploadAlt as upload,
  faCogs as cogsSolid,
  faExclamation as alert,
  faExclamationCircle as alertCircle,
  faExclamationTriangle as alertTriangle,
  faGift as giftSolid,
  faLock as lock,
  faPen as pen,
  faPhoneAlt as phone,
  faPrint as print,
  faQuestion as question,
  faQuestionCircle as questionCircle,
  faSignOutAlt as signOutSolid,
  faUmbrella as umbrellaSolid,
  faUsers as usersSolid,
  faTrash as trash,
  faVirus as virusSolid,
  faWrench as wrench,
} from '@fortawesome/pro-solid-svg-icons';

import {
  faCheckCircle as checkCircle,
  faFolder as folder,
  faSearch as search,
} from '@fortawesome/pro-regular-svg-icons';

import {
  faBallotCheck as ballotCheck,
  faCabinetFiling as cabinetFiling,
  faCircle as circle,
  faClipboardCheck as clipboardCheck,
  faCogs as cogs,
  faCreditCard as creditCard,
  faGift as gift,
  faPlus as plus,
  faSignOutAlt as signOut,
  faTimes as close,
  faTachometer as tachometer,
  faUmbrella as umbrella,
  faUser as user,
  faUsers as users,
  faVirus as virus,
} from '@fortawesome/pro-light-svg-icons';

import caretDown from '../../assets/icons/caret-down.svg';
import eyeClosed from '../../assets/icons/eye-closed.svg';
import eyeOpen from '../../assets/icons/eye-open.svg';

import getAttributes from '../../util/getAttributes';

const FAIcons = Object.freeze({
  alert,
  alertCircle,
  alertTriangle,
  analytics,
  arrowLeft,
  arrowRight,
  ballotCheck,
  ballotCheckSolid,
  ban,
  bell,
  building,
  cabinetFiling,
  cabinetFilingSolid,
  calendar,
  check,
  checkCircle,
  checkCircleSolid,
  circle,
  clipboardCheck,
  clipboardCheckSolid,
  clock,
  close,
  cogs,
  cogsSolid,
  creditCard,
  creditCardSolid,
  ear,
  folder,
  gift,
  giftSolid,
  lock,
  pen,
  phone,
  plus,
  print,
  question,
  questionCircle,
  search,
  signOut,
  signOutSolid,
  tachometer,
  tachometerSolid,
  trash,
  umbrella,
  umbrellaSolid,
  upload,
  user,
  userSolid,
  users,
  usersSolid,
  virus,
  virusSolid,
  wrench,
});

Object.values(FAIcons).forEach((icon) => {
  library.add(icon);
});

const CustomIcons = Object.freeze({
  caretDown,
  eyeClosed,
  eyeOpen,
});

export const Icons = Object.freeze(
  [...Object.keys(FAIcons), ...Object.keys(CustomIcons)].sort().reduce(
    (obj, val) => ({
      ...obj,
      [snakeCase(val).toUpperCase()]: val,
    }),
    {}
  )
);

export const isValidIcon = (value) =>
  Object.values(Icons).includes(camelCase(value));

const isCustomIcon = (iconName) => Object.keys(CustomIcons).includes(iconName);

const getComponent = (iconName) =>
  isCustomIcon(iconName) ? CustomIcons[iconName] : FontAwesomeIcon;

/**
 * `import { BaseIcon } from '@bambee/pollen';`
 *
 * An SVG icon component pulling from Bambee's predefined icon set. Icons are
 * designed to be styled according to the contextual size and font color, e.g.
 * if you want a red, 40px icon, make sure it or its containing element is
 * styled with `color: red; font-size: 40px;`.
 */
export default {
  props: {
    /** The name of the icon to use. */
    icon: {
      type: String,
      required: true,
      validator: isValidIcon,
    },
  },
  getAttributes,
  getComponent,
  getIcon: (icon) => FAIcons[camelCase(icon)] || null,
  isCustomIcon,
};
</script>

<style>
.base-icon--custom {
  height: 1em;
  width: 1em;
}
</style>
