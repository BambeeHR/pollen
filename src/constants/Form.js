const Form = Object.freeze({
  Variants: {
    STANDARD: 'standard',
    RAISED: 'raised',
    GHOST: 'ghost',
    GHOST_INVERTED: 'ghost-inverted',
  },
  Sizes: {
    DENSE: 'dense',
    NORMAL: 'normal',
    LARGE: 'large',
  },
});

export default Form;
