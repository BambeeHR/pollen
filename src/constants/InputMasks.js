/**
 * Commonly used configuration objects for use with Cleave. See:
 * https://github.com/nosir/cleave.js/blob/master/doc/options.md
 */
const InputMasks = Object.freeze({
  CURRENCY: {
    prefix: '$ ',
    numeral: true,
    numeralThousandsGroupStyle: 'thousand',
    noImmediatePrefix: true,
  },

  FEIN: {
    delimiters: ['-'],
    blocks: [2, 7],
    numericOnly: true,
  },

  PHONE: {
    delimiters: ['(', ') ', '-'],
    blocks: [0, 3, 3, 4],
    numericOnly: true,
  },

  YEAR: {
    blocks: [4],
    numericOnly: true,
  },

  ZIP_CODE: {
    blocks: [5],
    numericOnly: true,
  },
});

export default InputMasks;
