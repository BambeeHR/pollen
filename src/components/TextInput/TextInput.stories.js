import capitalize from 'lodash/capitalize';
import { action } from '@storybook/addon-actions';
import { boolean, select, text } from '@storybook/addon-knobs';
import TextInput from './TextInput.vue';
import Form from '../../constants/Form';
import TypeOverline from '../TypeOverline/TypeOverline.vue';
import { Icons } from '../BaseIcon/BaseIcon.vue';
import '../../assets/tailwind.css';

export default {
  title: 'Inputs/TextInput',
  component: TextInput,
};

const { Sizes, Variants } = Form;

export const Gallery = () => ({
  components: { TextInput, TypeOverline },
  data() {
    return {
      sizes: Sizes,
      variants: Variants,
    };
  },
  methods: {
    capitalize,
    handleInput: action('input'),
  },
  template: `
    <div>
      <TypeOverline tag="h1" variant="large">Regular Inputs</TypeOverline>
      <div class="flex flex-wrap -mx-8 -mt-8">
        <div
          v-for="variant in variants"
          :key="variant"
          :class="['w-56', (variant === 'ghost-inverted' ? 'm-4 p-4 bg-gray-0 text-white' : 'm-8')]"
        >
          <TypeOverline tag="h2" class="mb-2" variant="large">{{ variant }}</TypeOverline>
          <div v-for= "size in sizes" :key="'regular-' + variant + '-' + size" class="mb-2">
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Empty</TypeOverline>
            <TextInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Filled</TypeOverline>
            <TextInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              :value="capitalize(size)"
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Invalid</TypeOverline>
            <TextInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              invalid
              @input="handleInput"
            />            
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Disabled</TypeOverline>
            <TextInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              disabled
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Disabled, Filled</TypeOverline>
            <TextInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              disabled
              :value="capitalize(size)"
              @input="handleInput"
            />
          </div>
        </div>
      </div>
    </div>
  `,
});

export const WithKnobs = () => {
  const iconOptions = ['', ...Object.values(Icons)];
  return {
    components: { TextInput },
    props: {
      variant: {
        default: select('Variant', Object.values(Variants), Variants.STANDARD),
      },
      size: {
        default: select('Size', Object.values(Sizes), Sizes.NORMAL),
      },
      label: {
        default: text('Label', 'First name'),
      },
      placeholder: {
        default: text('Placeholder', 'Joe Smith'),
      },
      error: {
        default: text('Error', 'Something went wrong'),
      },
      preIcon: {
        default: select('Pre-Icon', iconOptions, Icons.SEARCH),
      },
      postIcon: {
        default: select('Post-Icon', iconOptions, Icons.LOCK),
      },
      prefix: {
        default: text('Prefix', '$'),
      },
      invalid: {
        default: boolean('Invalid', false),
      },
      showReset: {
        default: boolean('Show Reset', true),
      },
    },
    methods: {
      handleInput: action('input'),
    },
    template: `
      <div class="p-8" :class="variant === 'ghost-inverted' && 'bg-gray-0'">
        <TextInput
          :label="label"
          :placeholder="placeholder"
          :variant="variant"
          :size="size"
          :error="error"
          :pre-icon="preIcon"
          :post-icon="postIcon"
          :prefix="prefix"
          :invalid="invalid"
          :show-reset="showReset"
          @input="handleInput"
        />
      </div>
    `,
  };
};
