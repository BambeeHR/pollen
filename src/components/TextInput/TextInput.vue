<template>
  <FormField
    :id="id"
    :variant="variant"
    :size="size"
    :label="label"
    :error="error"
    :invalid="invalid"
    :disabled="disabled"
    :focused="hasFocus"
    class="text-input"
    :class="[
      `text-input__${variant}`,
      `text-input__${size}`,
      {
        'text-input__invalid': invalid || error,
        'text-input__disabled': disabled,
      },
    ]"
  >
    <BaseIcon
      v-if="preIcon"
      :icon="preIcon"
      class="text-input--icon text-input--icon__pre"
    />
    <span v-if="prefix" class="text-input--prefix">{{ prefix }}</span>
    <input
      :id="id"
      ref="input"
      :value="value"
      v-bind="$attrs"
      class="text-input--field"
      :aria-invalid="invalid"
      :aria-describedby="error ? '${id}-error' : false"
      :disabled="disabled"
      @input="handleInput"
      @focus="hasFocus = true"
      @blur="hasFocus = false"
    />
    <BaseIcon
      v-if="postIcon"
      :icon="postIcon"
      class="text-input--icon text-input--icon__post"
    />
    <IconButton
      v-if="showReset"
      :variant="$options.Button.Variants.SECONDARY"
      :size="$options.Button.Sizes.SMALL"
      class="text-input--reset text-input--icon"
      type="button"
      :icon="$options.Icons.CLOSE"
      title="Clear"
      flat
      :disabled="disabled"
      @click.prevent="handleReset"
    />
    <slot />
  </FormField>
</template>

<script>
import shortid from 'shortid';
import BaseIcon, { Icons, isValidIcon } from '../BaseIcon/BaseIcon.vue';
import Button from '../../constants/Button';
import Form from '../../constants/Form';
import FormField from '../internal/forms/FormField.vue';
import IconButton from '../IconButton/IconButton.vue';

const { Sizes, Variants } = Form;

export default {
  Button,
  Icons,
  components: {
    BaseIcon,
    FormField,
    IconButton,
  },
  inheritAttrs: false,
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    error: {
      type: String,
      default: null,
    },
    id: {
      type: String,
      default: shortid.generate,
    },
    invalid: {
      type: Boolean,
      default: false,
    },
    label: {
      type: String,
      default: null,
    },
    postIcon: {
      type: String,
      default: null,
      validator: (value) => !value || isValidIcon(value),
    },
    preIcon: {
      type: String,
      default: null,
      validator: (value) => !value || isValidIcon(value),
    },
    prefix: {
      type: String,
      default: null,
    },
    showReset: {
      type: Boolean,
      default: false,
    },
    size: {
      type: String,
      default: Sizes.NORMAL,
      validator: (value) => Object.values(Sizes).includes(value),
    },
    value: {
      type: String,
      default: '',
    },
    variant: {
      type: String,
      default: Variants.STANDARD,
      validator: (value) => Object.values(Variants).includes(value),
    },
  },
  data() {
    return {
      hasFocus: false,
    };
  },
  methods: {
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
    handleReset() {
      this.$refs.input.value = '';
      this.$emit('input', '');
    },
  },
};
</script>

<style scoped>
.text-input--prefix {
  @apply select-none text-gray-3;
}

.text-input--field {
  @apply appearance-none bg-transparent border-0 flex-grow w-full;
  font-size: 1em;
}

.text-input--field::placeholder {
  @apply text-gray-4;
}

.text-input--field:focus {
  @apply outline-none;
}

.text-input--icon,
.text-input--reset.icon-button {
  @apply text-gray-4;
}

.text-input--reset {
  @apply -mr-2;
}

.text-input__ghost-inverted .text-input--field,
.text-input__ghost-inverted .text-input--icon,
.text-input__ghost-inverted .text-input--prefix {
  @apply text-white;
}

.text-input__dense .text-input--icon {
  @apply text-14;
}

.text-input__normal .text-input--icon {
  @apply text-16;
}

.text-input__large .text-input--icon {
  @apply text-19;
}

.text-input__dense .text-input--icon__pre {
  @apply mr-3;
}

.text-input__normal .text-input--icon__pre {
  @apply mr-4;
}

.text-input__large .text-input--icon__pre {
  @apply mr-4;
}

.text-input__dense .text-input--prefix {
  @apply text-14;
  margin-right: 0.375rem;
}

.text-input__normal .text-input--prefix {
  @apply text-16;
  margin-right: 0.375rem;
}

.text-input__large .text-input--prefix {
  @apply text-18;
  @apply mr-2;
}

.text-input__dense .text-input--reset {
  @apply ml-2;
}

.text-input__normal .text-input--reset,
.text-input__large .text-input--reset {
  @apply ml-3;
}

.text-input__dense >>> .field-container {
  @apply h-8 px-3;
}

.text-input__normal >>> .field-container {
  @apply h-10 px-4;
}

.text-input__large >>> .field-container {
  @apply h-14 px-4;
}

.text-input__raised.text-input__invalid .text-input--field,
.text-input__raised.text-input__invalid .text-input--field::placeholder,
.text-input__ghost.text-input__invalid .text-input--field,
.text-input__ghost.text-input__invalid .text-input--field::placeholder {
  @apply text-error;
}

.text-input__ghost-inverted.text-input__invalid .text-input--field,
.text-input__ghost-inverted.text-input__invalid
  .text-input--field::placeholder {
  @apply text-error-light-ghost;
}
</style>
