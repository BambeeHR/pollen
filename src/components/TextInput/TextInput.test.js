import { mount } from '@vue/test-utils';
import Form from '../../constants/Form';
import InputMasks from '../../constants/InputMasks';
import TextInput from './TextInput.vue';
import { Icons } from '../BaseIcon/BaseIcon.vue';

const testIcon = Icons.LOCK;
const { Sizes, Variants } = Form;

describe('TextInput', () => {
  test('renders correctly', () => {
    const wrapper = mount(TextInput, {
      propsData: {
        variant: Variants.RAISED,
        size: Sizes.DENSE,
        preIcon: testIcon,
        postIcon: testIcon,
        error: 'There was an error',
        prefix: '$',
        label: 'Test',
        id: 'test',
        value: 'Hello world!',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders pre and post icons', () => {
    const wrapper = mount(TextInput, {
      propsData: {
        preIcon: testIcon,
        postIcon: testIcon,
      },
    });
    expect(wrapper.findAll('.base-icon').length).toBe(2);
  });

  test('transparently passes through classes and attributes', () => {
    const wrapper = mount(TextInput, {
      attrs: {
        id: 'test',
        type: 'email',
        disabled: true,
      },
    });
    const inputEl = wrapper.find('input').element;
    expect(inputEl.id).toBe('test');
    expect(inputEl.type).toBe('email');
    expect(inputEl.disabled).toBe(true);
  });

  test('emits an input event event', () => {
    const onInput = jest.fn();
    const wrapper = mount(TextInput, {
      listeners: {
        input: onInput,
      },
    });
    wrapper.find('input').setValue('hello world');
    expect(onInput).toHaveBeenCalledWith('hello world');
  });

  test('does not emit an input event when disabled', () => {
    const onInput = jest.fn();
    const wrapper = mount(TextInput, {
      listeners: {
        input: onInput,
      },
      propsData: {
        disabled: true,
      },
    });
    wrapper.find('input').setValue('hello world');
    expect(onInput).not.toHaveBeenCalledWith('hello world');
  });

  test('masks inputs', () => {
    const onInput = jest.fn();
    const wrapper = mount(TextInput, {
      listeners: {
        input: onInput,
      },
      propsData: {
        mask: InputMasks.PHONE,
      },
    });
    const input = wrapper.find('input');
    input.setValue('1234567890');
    expect(onInput).toHaveBeenCalledWith('1234567890');
    expect(wrapper.find('input').element.value).toBe('(123) 456-7890');
  });
});
