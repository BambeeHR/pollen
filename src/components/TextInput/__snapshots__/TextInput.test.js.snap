// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders correctly 1`] = `
<label
  class="form-field text-input text-input__raised text-input__dense text-input__invalid"
  for="test"
>
  <div
    class="field-label field-label__raised field-label__dense field-label__invalid"
  >
    
    Test
  
  </div>
   
  <div
    class="field-container field-container__raised field-container__dense field-container__invalid"
  >
    <svg
      aria-hidden="true"
      class="base-icon svg-inline--fa fa-lock fa-w-14 text-input--icon text-input--icon__pre"
      data-icon="lock"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
        fill="currentColor"
      />
    </svg>
     
    <span
      class="text-input--prefix"
    >
      $
    </span>
     
    <input
      aria-describedby="\${id}-error"
      class="text-input--field"
      id="test"
    />
     
    <svg
      aria-hidden="true"
      class="base-icon svg-inline--fa fa-lock fa-w-14 text-input--icon text-input--icon__post"
      data-icon="lock"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
        fill="currentColor"
      />
    </svg>
     
    <!---->
  </div>
   
  <div
    class="field-error field-error__raised"
    id="test-error"
    role="alert"
  >
    
    There was an error
  
  </div>
</label>
`;
